### ISSUE TITLE
Make sokol context based instead of being static globals

### ISSUE THREAD

**Issue by eloraiby** [floooh/sokol#82]
This allow sokol to have multiple contexts: Useful in case of multi-threaded programs, where the context on some platforms must be held in the thread that created it (e.g. Windows).

Are there any design reasons preventing this ?

**Comment by floooh** [comment]
sokol does have multiple-context-support, see the multi-window-example here, where each window has its own GL context, and sokol-gfx manages different contexts:

https://github.com/floooh/sokol-samples/blob/master/glfw/multiwindow-glfw.c

This is only tested with OpenGL though, where the concept of shared contexts exists (and is supported by GLFW).

You can create different context on the other APIs too, but this only sorts resources into context-specific pools.

Other then that, sokol-gfx itself should be thread-agnostic. You should be able to run it off the main-thread as long as all calls happen on the same thread, and the 3D context was created in the same thread (e.g. I think when using Metal on iOS with MTKKit, all rendering happens in a separate thread).

**Comment by floooh** [comment]
Yep, I see your point and have been thinking about this quite a bit in the beginning. I've decided against it though, because I don't think it's as easy as having a context/state pointer that's passed to functions. Except for GL you would still only be able to create a single global sokol_gfx state, and even with GL the multiple GL contexts need to live on the same thread).

OpenGL has the concept of multiple contexts per applications, but its not very well supported, adds complexity (e.g. shared resources between contexts), and AFAIK it's not simply possible to move GL context to different threads. It gets even murkier on D3D and Metal. AFAIK on Metal you can have multiple devices, *but* AFAIK that's intended to access multiple GPUs (so one device per app per GPU, not multiple devices accessing the same GPU from the same app). D3D11 is most likely the same, or if not it is most like not well tested.

Spreading rendering work over multiple threads in modern 3D APIs is basically having a single device on one thread, and multiple threads that generate command lists, when are then enqueued per frame on the 'device thread'. This will be a topic for sokol-gfx 2.0, but at the cost of removing support for GL and D3D11 (and instead use Vulkan and maybe D3D12).

**Comment by septag** [comment]
It's a great feature if we could have a multi-threaded API, but as you said, the native support is lacking within the current APIs, unless we leave it to another lib (sokol_gfx 2)
On the current sokol_gfx, we could do this ourselves and make another layer on top of sokol functions (contexts) that queue and merge the commands on the device thread .. because at the current state, I think sokol_gfx is pretty good in terms of low-levelness and complexity. 
I think bgfx does the same thing, but the simple/thin-layer design of sokol_gfx is what makes it special :)

It's off topic, but the main feature I really appreciate right now is support for compute-shaders. I know there are some APIs that doesn't support it (webgl, gles, gl3.3), but almost all PC hardware has support for compute-shaders nowadays and if you could add GL4 and GLES3.1 support, we could have compute-shaders for many platforms and devices. This leaves out webgl, which I guess webgl2 will be supporting it in the near future.

**Comment by floooh** [comment]
Ok noted :) AFAIK there's currently work underway to add compute shader support to WebGL2, but I have no idea if this will see any kind of adoption before WebGPU hits (which will have compute support built in anyway).

I also thought about adding a command queue to sokol-gfx. I had something similar implemented in Oryol when I experimented with GL uniform buffers, basically have one big GL uniform buffer and record all uniform updates in there in a "recording pass", and at the same time write the rendering commands to a command buffer along with offsets into the uniform buffer. This added quite a bit of code complexity without performance benefit, so in the end I dropped it again, and didn't pursue this idea further when starting with sokol-gfx.

I think the main reason BGFX does something similar is because it re-orders draw calls before rendering, to minimize the number of state switches. sokol-gfx sort of expects that the draw calls are already issued in an optimal order (although with Metal and to some extent also D3D11 this isn't super-important anymore).

So, IMHO exposing a command buffers mainly makes sense if the underlying API supports it too (Metal, D3D12 and Vulkan).

**Comment by eloraiby** [comment]
can't we pass some function to sokol to set the current context to current. something similar to `glfwMakeContextCurrent` ? that would solve the multithreading issue, no ?

**Comment by floooh** [comment]
No, I think that would at least need some sort of thread-awareness (to store context-specific thread-local data for the resource pools and state cache) in sokol. Also, for the GL backend, sokol itself doesn't have any knowledge of the GL context itself, since that's managed by the outside world (e.g. GLFW or sokol_app.h).

Last time I dabbled with multiple GL contexts and threads it was quite a compatibility nightmare, I only tried this on Windows, and it worked on one machine, but not another with a different graphics card and GL driver. I'd really like to avoid this mess ;) The right way to do multithreaded-rendering is what Vulkan, DX12 and Metal have arrived at (create command lists on threads, but enqueue them each frame on a single "render thread" for rendering).
